### PR TITLE
[merge jam] port 49200

### PR DESCRIPTION
**Pull request by waynew**
### What does this PR do?

Ports #49200

### What issues does this PR fix or reference?

N/A

### New Behavior

user/passwd functionality added to couchdb returner

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

File: salt/returners/couchdb_return.py [hunk 1]
```diff
@@ -74,6 +78,22 @@
...
+    boltons_lib = False
+
+# boltons lib for redact_pws and minimum_return
+try:
+    from boltons.iterutils import remap
+
+    boltons_lib = True
+except ImportError:
+    boltons_lib = False
+
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:96]: Is there a reason this try statement is in here twice?

File: salt/returners/couchdb_return.py [hunk 2]
```diff
@@ -122,14 +193,19 @@ def _generate_doc(ret):
...
     """
     Makes a HTTP request. Returns the JSON parse, or an obj with an error.
     """
     opener = _build_opener(_HTTPHandler)
     request = _Request(url, data=_data)
     if content_type:
         request.add_header("Content-Type", content_type)
+    if user and passwd:
+        auth_encode = "{0}:{1}".format(user, passwd).encode("base64")[:-1]
+        auth_basic = "Basic {0}".format(auth_encode)
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:206]: Pretty sure these format statements will need to have the numbers removed. So just `{}`

File: salt/returners/couchdb_return.py [hunk 3]
```diff
@@ -138,6 +214,30 @@ def _request(method, url, content_type=None, _data=None):
     return salt.utils.json.loads(handler.read())
 
 
+def _generate_event_doc(event):
+    """
+    Create a object that will be saved into the database based in
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:219]: Create **_an_** object that will be saved into the database based **_on_**

File: salt/returners/couchdb_return.py [hunk 4]
```diff
@@ -146,22 +246,53 @@ def returner(ret):
...
 
         # Confirm that the response back was simple 'ok': true.
         if "ok" not in _response or _response["ok"] is not True:
-            log.error("Unable to create database '%s'", options["db"])
-            log.error("Nothing logged! Lost data.")
+            log.error(
+                'Nothing logged! Lost data. Unable to create database "{0}"'.format(
+                    options["db"]
+                )
+            )
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:271]: You can use newer formatting strings with log messages:
```
log.error(
    'Nothing logged! Lost data. Unable to create database "%s"', options["db"]
)

File: salt/returners/couchdb_return.py [hunk 5]
```diff
@@ -146,22 +246,53 @@ def returner(ret):
...
+            log.debug("_response object is: {0}".format(_response))
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:272]:
```
log.debug("_response object is: %s", _response)
```

File: salt/returners/couchdb_return.py [hunk 6]
```diff
@@ -146,22 +246,53 @@ def returner(ret):
...
+
+        # remove all return values starting with '__pub' if options['minimum_return'] is True
+        if options["minimum_return"]:
+            ret_remapped = remap(ret_remap_pws, visit=_minimize_return)
+        else:
+            ret_remapped = ret_remap_pws
+    else:
+        log.info(
+            "boltons library not installed. Cannot redact passwords or minimize the return. pip install boltons. https://github.com/mahmoud/boltons."
+        )
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:291]: Should this maybe be a warning?

File: salt/returners/couchdb_return.py [hunk 7]
```diff
@@ -173,8 +304,70 @@ def returner(ret):
...
+    Requires that configuration be enabled via 'event_return'
+    option in master config.
+
+    Example:
+
+    event_return:
+      - couchdb
+
+    """
+    log.debug("events data is: {}".format(events))
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:324]:
```
log.debug("events data is: %s", events)
```

File: salt/returners/couchdb_return.py [hunk 8]
```diff
@@ -173,8 +304,70 @@ def returner(ret):
...
+
+    options = _get_options()
+
+    # Check to see if the database exists.
+    _response = _request("GET", options["url"] + "_all_dbs")
+    event_db = "{}-events".format(options["db"])
+    if event_db not in _response:
+
+        # Make a PUT request to create the database.
+        log.info('Creating database "{0}"'.format(event_db))
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:334]:
```
log.info('Creating database "%s"', event_db)
```

File: salt/returners/couchdb_return.py [hunk 9]
```diff
@@ -173,8 +304,70 @@ def returner(ret):
...
+            passwd=options["passwd"],
+        )
+
+        # Confirm that the response back was simple 'ok': true.
+        if "ok" not in _response or _response["ok"] is not True:
+            log.error(
+                'Nothing logged! Lost data. Unable to create database "{0}"'.format(
+                    event_db
+                )
+            )
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:348]:
```
log.error(
    'Nothing logged! Lost data. Unable to create database "%s"', event_db
)
```

File: salt/returners/couchdb_return.py [hunk 10]
```diff
@@ -173,8 +304,70 @@ def returner(ret):
...
+            return
+        log.info('Created database "{0}"'.format(event_db))
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:350]:
```
log.info('Created database "%s"', event_db)
```

File: salt/returners/couchdb_return.py [hunk 11]
```diff
@@ -173,8 +304,70 @@ def returner(ret):
...
+
+    for event in events:
+        # Call _generate_doc to get a dict object of the document we're going to shove into the database.
+        log.debug("event data is: {}".format(event))
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:354]:
```
log.debug("event data is: %s", event)
```

File: salt/returners/couchdb_return.py [hunk 12]
```diff
@@ -173,8 +304,70 @@ def returner(ret):
...
+            "application/json",
+            salt.utils.json.dumps(doc),
+        )
+        # Sanity check regarding the response..
+        if "ok" not in _response or _response["ok"] is not True:
+            log.error(
+                'Nothing logged! Lost data. Unable to create document: "{0}"'.format(
+                    _response
+                )
+            )
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:370]:
```
log.error(
    'Nothing logged! Lost data. Unable to create document: "%s"', _response
)
```

File: salt/returners/couchdb_return.py [hunk 13]
```diff
@@ -375,3 +567,21 @@ def save_minions(jid, minions, syndic_id=None):  # pylint: disable=unused-argume
...
+
+
+def get_load(jid):
+    """
+    Return the load data that marks the specified jid.
+    """
+    options = _get_options(ret=None)
+    _response = _request("GET", options["url"] + options["db"] + "/" + jid)
+    if "error" in _response:
+        log.error('Unable to get JID "{0}" : "{1}"'.format(jid, _response))
```
Comment (twangboy) [inline salt/returners/couchdb_return.py:585]:
```
log.error('Unable to get JID "%s" : "%s"', jid, _response)
```

**Review by twangboy — CHANGES_REQUESTED** [review]
Pretty sure this will need some tests as well

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**